content/blog/zardshard/2023-08-18_[gsoc_2023]_improving_icon-o-matic_final_report.md

```diff
@@ -0,0 +1,92 @@
++++
+type = "blog"
+title = "[GSoC 2023] Improving Icon-O-Matic Final Report"
+author = "Zardshard"
+date = "2023-08-18 11:21:50-04:00"
+tags = ["haiku", "software", "icon-o-matic", "gsoc", "gsoc2023"]
++++
+
+## What is Icon-O-Matic?
+
+![Screenshot of Icon-O-Matic](/files/blog/zardshard/icon_o_matic.png)
+
+There's a good chance that not everyone reading this article will know what Icon-O-Matic is, so I'll start by explaining what it is. Icon-O-Matic is a vector graphics editing program like Illustrator or Inkscape. It is specifically made to work with Haiku's custom HVIF vector graphics format. This format is similar to the SVG format, except optimized to be much, much smaller. The blog post ["500 Byte Images: The Haiku Vector Icon Format"](http://blog.leahhanson.us/post/recursecenter2016/haiku_icons.html) provides a more in-depth discussion for those interested.
+
+If you want to try Icon-O-Matic yourself, you will need to first install Haiku OS. Afterwards, you can launch it from the applications menu.
+
+## My goals
+
+The goals of my project were rather broad: to add features, refactor code, and fix bugs. I had, therefore, a lot of leeway to do what interested me. My early plans were detailed in a blog post entitled ["[GSoC 2023] Plans for improving Icon-O-Matic"](https://www.haiku-os.org/blog/zardshard/2023-05-05_gsoc_2023_plans_for_improving_icon-o-matic/). They changed rather significantly over the course of the project.
+
+## What I got done
+
+Below is a list of the things I got done during GSOC. For those interested in seeing the full list, it can be seen [here](https://review.haiku-os.org/q/owner:0azrune6%2540zard.anonaddy.com). GSoC ended for me on August 28, so commits before that date were submitted during GSoC.
+
+### Fixed memory leaks
+
+For those not familiar with it, a memory leak occurs when a program requests memory from the operating system and later forgets to tell the operating system that it's done with it. The symptoms of this problem develop slowly over time. As the user continues to use the program, it will slowly but surely take up more and more memory.
+
+One of my GSoC project ideas was to make a program to detect memory leaks. I had decided not to go with that. Later, I discovered that Haiku had a memory leak detector built-in! I wrote [a blog post](https://www.haiku-os.org/blog/zardshard/2023-05-23_how_to_find_memory_leaks/) detailing how to use it so that others knew of its existence and how to use it. I was excited to use this memory leak detector on Icon-O-Matic, so, that was one of the first things I did.
+
+And, indeed, I did find and fix many memory leaks. Some of these leaks were caused by system libraries. This means that some of the bugfixes helped every application leak less memory.
+
+For those interested, there remain memory leaks in Icon-O-Matic that I did not get to. Doubtless there are many memory leaks in nearly every application in Haiku OS.
```

Comment:
That is not very nice to say :sweat_smile: 

Most of the leaks were things that "leak" only once at the startup of the application, and so, they are not really leaks. There will certainly be some, but hopefully not that many leaks and not in every app?

Reply:
Ok, if leaks only at startup do not count as leaks, then, that statement is unjustified.